### ISSUE TITLE
After being changed to another string the first element of a list[str], pylance infers it as None

### ISSUE THREAD

**Issue by PatrickBourdon** [microsoft/pylance-release#992]
## Environment data
-   Windows-10-10.0.18362-SP0
-   python(3.9.2 (tags/v3.9.2:1a79785, Feb 19 2021, 13:44:55) [MSC v.1928 64 bit (AMD64)])
-   Visual Code Studio: 1.53.2
-   extension pylance: v2021.2.4
-   type checking mode: strict

## Foreword 
I have just finished to (strictly) type several thousands of python lines of code using pylance.
It's a great tool that enforces to make the code more clearer and maintainable.
I encountered some issues, some were bugs, some borderline cases.
I want to thank Erik and Jake for the reactivity and for the very clear answers.
This is probably the last issue, a bit surprising.

## Expected behaviour
No problem reported.

## Actual behaviour
A problem reported. See example below

## Code Snippet / Additional information
```python
y: list[str] = [ 'a', 'b', 'c' ]
y[0] = 'b'
y0 = y[0]
if y[0].lower() == 'a':
    ...
# Cannot access member "lower" for type "None"
#  Type "None" is unsupported

z = y
z0 = z[0]
if z[0].lower() == 'a':
    ...
# all is correct here
```

**Comment by erictraut** [comment]
Thanks for the bug report. At first glance, this looks like a bug in a feature I added in the latest release of Pylance: type narrowing for subscript (index) expressions. The type of the expression `y[0]` is apparently being narrowed to `None` when it shouldn't be. I'll need to investigate and debug.

**Comment by erictraut** [comment]
Yes, this was a bug in the assignment-based type narrowing for index expressions. This code was recently added (appeared in the last release of pylance), and it contained a bug. My unit test explicitly assigned `None`, so it didn't catch the problem. I've fixed the issue and beefed up the unit test.

@PatrickBourdon, it's gratifying to hear that you're finding pylance useful. I'm impressed that you got several thousand lines of code to pass strict type checking. That's not easy to do. My team has about 250K lines of Python code, and it took us about a year and a half to get it to pass under strict type checks. Now it's much easier to maintain.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.3.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202130-3-march-2021
